### PR TITLE
Perf logging & Incremental Compilation

### PR DESCRIPTION
**Pull request by alxhub**

### CODE REVIEW

File: packages/compiler-cli/src/ngcc/src/analysis/decoration_analyzer.ts [hunk 1]
```diff
@@ -6,6 +6,7 @@
  * found in the LICENSE file at https://angular.io/license
  */
 import {ConstantPool} from '@angular/compiler';
+import {NOOP_PERF_RECORDER} from '@angular/compiler-cli/src/ngtsc/perf';
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngcc/src/analysis/decoration_analyzer.ts]: I don't think this is used.

File: packages/compiler-cli/src/ngtsc/perf/src/tracking.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    this.log.push(msg);
+    return this.inSpan;
+  }
+
+  stop(span: number|false): void {
+    if (this.inSpan !== span) {
+      throw new Error(`Perf Error: stop(${span}) but not currently in that span`);
+    }
+    this.inSpan = this.spanStack.pop() || null;
+    this.log.push({
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/perf/src/tracking.ts]: May be use `makeLogMessage` for the `stop` method as well? In this case it might also be possible to combine log producing operation and storing the result in `this.log`.

File: packages/compiler-cli/src/ngtsc/perf/src/tracking.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    const msg = this.makeLogMessage(PerfLogEventType.SPAN_OPEN, name, node, category, detail);
+    this.log.push(msg);
+    return this.inSpan;
+  }
+
+  stop(span: number|false): void {
+    if (this.inSpan !== span) {
+      throw new Error(`Perf Error: stop(${span}) but not currently in that span`);
+    }
+    this.inSpan = this.spanStack.pop() || null;
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/perf/src/tracking.ts]: nit: I'd propose renaming `inSpan` field to something like `currentSpanId`. The `in` prefix might seem confusing and look like field type is boolean :)

File: packages/compiler-cli/src/ngcc/src/analysis/decoration_analyzer.ts [hunk 4]
```diff
@@ -90,7 +91,8 @@ export class DecorationAnalyzer {
         this.reflectionHost, this.evaluator, this.scopeRegistry, NOOP_DEFAULT_IMPORT_RECORDER,
         this.isCore),
     new InjectableDecoratorHandler(
-        this.reflectionHost, NOOP_DEFAULT_IMPORT_RECORDER, this.isCore, /* strictCtorDeps */ false),
+        this.reflectionHost, NOOP_DEFAULT_IMPORT_RECORDER, this.isCore,
+        /* strictCtorDeps */ false),
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngcc/src/analysis/decoration_analyzer.ts]: Looks like formatting issue..

File: packages/compiler-cli/src/ngtsc/incremental/src/state.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    for (const oldFile of oldProgram.getSourceFiles()) {
+      if (!oldFile.isDeclarationFile) {
+        oldFiles.add(oldFile);
+      }
+    }
+
+    // Look for files in the new program which haven't changed.
+    for (const newFile of newProgram.getSourceFiles()) {
+      if (oldFiles.has(newFile)) {
+        !newFile.isDeclarationFile && console.error(`U ${newFile.fileName}`);
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/incremental/src/state.ts]: Is this change intentional or left for debugging? If it's needed, why do we use `console.error`, but not `console.log`, since it seems informational mostly?

File: packages/compiler-cli/src/ngtsc/program.ts [hunk 6]
```diff
@@ -57,10 +59,18 @@ export class NgtscProgram implements api.Program {
...
+  private perfRecorder: PerfRecorder = NOOP_PERF_RECORDER;
+  private perfTracker: PerfTracker|null = null;
+  private incrementalState: IncrementalState;
 
   constructor(
       rootNames: ReadonlyArray<string>, private options: api.CompilerOptions,
       host: api.CompilerHost, oldProgram?: api.Program) {
+    if (shouldEnablePerfTracing(options)) {
+      this.perfTracker = PerfTracker.zeroedToNow();
+      this.perfRecorder = this.perfTracker;
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/program.ts]: Why do we need `perfTracker` and `perfRecorder` (since they also refer to the same instance)? (sorry if I missed it from the code or PR description)

File: packages/compiler-cli/src/transformers/api.ts [hunk 7]
```diff
@@ -205,6 +205,17 @@ export interface CompilerOptions extends ts.CompilerOptions {
   /** @internal */
   collectAllErrors?: boolean;
 
+  /** An option to enable ngtsc's internal performance tracing.
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/transformers/api.ts]: Formatting issue here (text should start next line) and at the end of comment block (`*/` should be at the next line).

**Review by AndrewKushnir — APPROVED** [review]
Looks great (just left a few nits) 👍 
One question/proposal: may be add some test for PerfTracker logic?

File: packages/compiler-cli/src/ngtsc/perf/src/tracking.ts (hunk 2 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/perf/src/tracking.ts]: Done - now the start/stop combo produces one message with a `duration`.

File: packages/compiler-cli/src/ngtsc/perf/src/tracking.ts (hunk 3 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/perf/src/tracking.ts]: I removed it :)

File: packages/compiler-cli/src/ngcc/src/analysis/decoration_analyzer.ts (hunk 4 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngcc/src/analysis/decoration_analyzer.ts]: Fixed.

File: packages/compiler-cli/src/ngtsc/incremental/src/state.ts (hunk 5 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/incremental/src/state.ts]: Removed :)

File: packages/compiler-cli/src/transformers/api.ts (hunk 7 above)
Comment (alxhub) [reply packages/compiler-cli/src/transformers/api.ts]: Fixed.

File: packages/compiler-cli/src/ngtsc/program.ts (hunk 6 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/program.ts]: They have different types.

`perfTracker` gives the full `PerfTracker` interface, and is `null` if perf tracking is not enabled.
`perfRecorder` only gives the `PerfRecorder` interface, and is always set (to the `NOOP_PERF_RECORDER` if tracing isn't enabled).

**Comment by alxhub** [comment]
As far as tests go, I'd rather wait until we stabilize the logging format before we start explicitly testing it.

File: packages/compiler-cli/src/ngtsc/incremental/src/state.ts [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    if (!this.metadata.has(sf)) {
+      // The file has no metadata from the previous or current compilations, so it must be emitted.
+      return false;
+    }
+
+    const meta = this.metadata.get(sf) !;
+
+    // Check if this file was explicitly marked as safe. This would only be done if every
+    // `DecoratorHandler` agreed that the file didn't depend on any other file's contents.
+    if (meta.safeToSkipEmitIfUnchanged) {
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/incremental/src/state.ts]: nit: we can probably just return `meta.safeToSkipEmitIfUnchanged` (without affecting readability), like:
```
 return meta.safeToSkipEmitIfUnchanged;
```

File: packages/compiler-cli/src/ngtsc/incremental/src/state.ts [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+/**
+ * Accumulates state between compilations.
+ */
+export class IncrementalState {
+  private unchangedFiles: Set<ts.SourceFile>;
+  private metadata: Map<ts.SourceFile, FileMetadata>;
+
+  private constructor(
+      unchangedFiles: Set<ts.SourceFile>, metadata: Map<ts.SourceFile, FileMetadata>) {
+    this.unchangedFiles = unchangedFiles;
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/incremental/src/state.ts]: It looks like assignments in constructor body can be omitted (it'd be taken care of by TS), since we just assign respective values without any changes?

File: packages/compiler-cli/src/ngtsc/perf/src/tracking.ts [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    const msg = this.openSpans.get(span) !;
+    this.openSpans.delete(span);
+
+    // Update the message to add a duration.
+    msg.duration = timeSinceInMicros(this.zeroTime) - msg.stamp;
+    this.log.push(msg);
+  }
+
+  private makeLogMessage(
+      type: PerfLogEventType, name: string, node: ts.SourceFile|ts.Declaration|undefined,
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/perf/src/tracking.ts]: Can we make `node`, `category`, `detail` and `span` optional (instead of `|undefined`)?

File: packages/compiler-cli/src/ngtsc/perf/src/tracking.ts [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        const name = ts.getNameOfDeclaration(node);
+        if (name !== undefined && ts.isIdentifier(name)) {
+          msg.declaration = name.text;
+        }
+      }
+    }
+    return msg;
+  }
+
+  asJson(): unknown { return this.log; }
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/perf/src/tracking.ts]: `this.log`'s type is defined as `PerfLogEvent[]`, can we set it as return value type instead of `unknown`?

File: packages/compiler-cli/src/ngtsc/program.ts (hunk 6 above)
Comment (AndrewKushnir) [reply packages/compiler-cli/src/ngtsc/program.ts]: Ok, thanks for additional info!

**Review by AndrewKushnir — APPROVED** [review]
Looks great (just left a few minor nits)! 👍

File: packages/compiler-cli/src/ngtsc/perf/src/tracking.ts (hunk 10 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/perf/src/tracking.ts]: I want them to be explicitly passed (just like how `{foo?: string}` and `{foo: string|undefined}` are not the same thing in TS)

File: packages/compiler-cli/src/ngtsc/perf/src/tracking.ts (hunk 11 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/perf/src/tracking.ts]: No, because `PerfLogEvent` is internal. The caller of `asJson()` gets a blob.

File: packages/compiler-cli/src/ngtsc/incremental/src/state.ts (hunk 8 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/incremental/src/state.ts]: I actually disagree this doesn't affect readability. This is very critical code - subtly incorrect code can be emitted if this isn't done correctly. So I want skipping emit to be _explicitly opt-in_, and the default path will obviously be the explicit `return false` if the conditions for skipping emit are met.

Comment (AndrewKushnir) [reply packages/compiler-cli/src/ngtsc/incremental/src/state.ts]: Ok, makes sense, thanks!

**Comment by AndrewKushnir** [comment]
@alxhub thanks for answering questions and for additional changes! Everything looks great, I have no other questions/comments. 👍
